text/0000-pattern-3.md

```diff
@@ -0,0 +1,1653 @@
+}
+```
+
+We allow a hay to customize the `Index` type. While `str`, `[T]` and `OsStr` all  use `usize` as
+the index, we do want the Pattern API to support other linear structures like `LinkedList<T>`,
+where a cursor/pointer would be more suitable for allowing sub-linear splitting.
+
+```
+start_index() = 0   next_index(2) = 6
+        |         +-------------------+
+        v         ^                   v
+        0    1    2    3    4    5    6    7
+        +----+----+----+----+----+----+----+
+        | 48 | 69 | f0   9f   8c   8d | 21 |
+        +----+----+----+----+----+----+----+
+        0    1    2    3    4    5    6    7
+             ^    v                        ^
+             +----+                        |
+        prev_index(2) = 1            end_index() = 7
+```
+
+### Haystack
+
+A `Haystack` is any linear structure which we can do pattern matching on,
+and can be sliced or split so they could be returned from the `matches()` and `split()` iterators.
+
+Haystack is implemented on the reference or collection itself e.g. `&[T]`, `&mut [T]` and `Vec<T>`.
+A hay can *borrowed* from a haystack.
+
+```rust
+pub trait Haystack: Deref<Target: Hay> + Sized {
+    fn empty() -> Self;
+    unsafe fn split_around(self, range: Range<Self::Target::Index>) -> [Self; 3];
+    unsafe fn slice_unchecked(self, range: Range<Self::Target::Index>) -> Self;
+
+    fn restore_range(
+        &self,
+        original: Range<Self::Target::Index>,
+        parent: Range<Self::Target::Index>,
+    ) -> Range<Self::Target::Index>;
+}
+// we assume either RFC 2089 (issue #44491) or RFC 2289 is implemented.
+// for simplicity we're ignoring issue #38078 which forces us to write `<Self::Target as Hay>::Index`.
+```
+
+> We assume either Implied Bounds ([RFC 2089] / [issue 44491]) is implemented (thus fixing
+> [issue 20671]), or Associated Type Bounds ([RFC 2289]) has been accepted and implemented.
+>
+> For simplicity we are ignoring [issue 38078],
+> which forces us to write `<Self::Target as Hay>::Index` instead of `Self::Index`.
+
+The `self.restore_range(original, parent)` method is implemented to solve:
+
+* given haystacks `a` and `b`
+* given `a = b[original]` and `self = a[parent]`
+* find `range` such that `self == b[original][parent] == b[range]`
+
+This method is used to recover the original range in functions like `find()` and `match_indices()`.
+It is usually just implemented as `(original.start + parent.start)..(original.start + parent.end)`.
+
+When an index is based on a pointer, splitting a haystack will *invalidate* those pointers.
+However, a pointer is persisted with slicing, so they could implement this method simply as
+`self.start_index()..self.end_index()`.
+
+### Shared haystack
+
+A `SharedHaystack` is a marker sub-trait which tells the compiler this haystack can cheaply be
+cheaply cloned (i.e. shared), e.g. a `&H` or `Rc<H>`. Implementing this trait alters some behavior
+of the `Span` structure discussed next section.
```

Comment:
I'm somewhat uncomfortable by the use of specialization to modify the behaviour of `Span` instead of just providing more optimized functions. Admittedly, this changed behaviour seems hard to directly observe, since `into_parts` is only available on shared spans. This definitely isn't a big deal.